components/webcompat/webcompat_exceptions_service.cc

```diff
@@ -0,0 +1,192 @@
+        {"language", BRAVE_WEBCOMPAT_LANGUAGE},
+        {"media-devices", BRAVE_WEBCOMPAT_MEDIA_DEVICES},
+        {"plugins", BRAVE_WEBCOMPAT_PLUGINS},
+        {"screen", BRAVE_WEBCOMPAT_SCREEN},
+        {"speech-synthesis", BRAVE_WEBCOMPAT_SPEECH_SYNTHESIS},
+        {"usb-device-serial-number", BRAVE_WEBCOMPAT_USB_DEVICE_SERIAL_NUMBER},
+        {"user-agent", BRAVE_WEBCOMPAT_USER_AGENT},
+        {"webgl", BRAVE_WEBCOMPAT_WEBGL},
+        {"webgl2", BRAVE_WEBCOMPAT_WEBGL2},
+        {"websockets-pool", BRAVE_WEBCOMPAT_WEB_SOCKETS_POOL},
+    });
+
+WebcompatExceptionsService* singleton = nullptr;
+
+}  // namespace
+
+WebcompatExceptionsService::WebcompatExceptionsService(
+    LocalDataFilesService* local_data_files_service)
+    : LocalDataFilesObserver(local_data_files_service) {}
+
+void WebcompatExceptionsService::LoadWebcompatExceptions(
+    const base::FilePath& install_dir) {
+  base::FilePath txt_file_path =
+      install_dir.AppendASCII(WEBCOMPAT_EXCEPTIONS_JSON_FILE_VERSION)
+          .AppendASCII(WEBCOMPAT_EXCEPTIONS_JSON_FILE);
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
+      base::BindOnce(&brave_component_updater::GetDATFileAsString,
+                     txt_file_path),
+      base::BindOnce(&WebcompatExceptionsService::OnJsonFileDataReady,
+                     weak_factory_.GetWeakPtr()));
+}
+
+bool WebcompatExceptionsService::AddRule(const ContentSettingsPattern& pattern,
+                                         const std::string& exception_string) {
+  const auto it = kWebcompatNamesToType.find(exception_string);
+  if (it != kWebcompatNamesToType.end()) {
+    const auto webcompat_type = it->second;
+    if (!patterns_by_webcompat_type_.contains(webcompat_type)) {
+      patterns_by_webcompat_type_[webcompat_type] = kEmptyPatternVector;
+    }
+    patterns_by_webcompat_type_[webcompat_type].push_back(pattern);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+void WebcompatExceptionsService::AddRules(
+    const base::Value::List& include_strings,
+    const base::Value::Dict& rule_dict) {
+  const base::Value* exceptions = rule_dict.Find(kExceptions);
+  if (exceptions->is_list()) {
+    for (const base::Value& include_string : include_strings) {
+      const auto pattern =
+          ContentSettingsPattern::FromString(include_string.GetString());
+      for (const base::Value& exception : exceptions->GetList()) {
+        const bool success = AddRule(pattern, exception.GetString());
+        if (!success) {
+          DLOG(ERROR) << "Unrecognized webcompat exception "
+                      << exception.GetString();
+        }
+      }
+    }
+  } else {
+    DLOG(ERROR) << "Malformed exceptions list in "
+                << WEBCOMPAT_EXCEPTIONS_JSON_FILE;
+  }
+}
+
+const std::vector<ContentSettingsPattern>&
+WebcompatExceptionsService::GetPatterns(ContentSettingsType webcompat_type) {
+  if (patterns_by_webcompat_type_.contains(webcompat_type)) {
```

Comment:
this is a double lookup, why not do `find`

Reply:
fixed